protos/icmp/icmp.go

```diff
@@ -0,0 +1,329 @@
+	}
+
+	icmp.transactions = common.NewCacheWithRemovalListener(
+		protos.DefaultTransactionExpiration,
+		protos.DefaultTransactionHashSize,
+		removalListener)
+
+	icmp.transactions.StartJanitor(protos.DefaultTransactionExpiration)
+
+	icmp.results = results
+
+	return icmp, nil
+}
+
+func (icmp *Icmp) InitDefaults() {
+	icmp.SendRequest = false
+	icmp.SendResponse = false
+}
+
+func (icmp *Icmp) SetFromConfig(config config.Icmp) (err error) {
+	if config.Send_request != nil {
+		icmp.SendRequest = *config.Send_request
+	}
+	if config.Send_response != nil {
+		icmp.SendResponse = *config.Send_response
+	}
+
+	return nil
+}
+
+func (icmp *Icmp) ProcessICMPv4(icmp4 *layers.ICMPv4, pkt *protos.Packet) {
+	typ := uint8(icmp4.TypeCode >> 8)
+	code := uint8(icmp4.TypeCode)
+	id, seq := extractTrackingData(4, typ, &icmp4.BaseLayer)
+	tuple := NewIcmpTuple(4, pkt.Tuple.Src_ip, pkt.Tuple.Dst_ip, id, seq)
+	msg := IcmpMessage{
+		Ts:     pkt.Ts,
+		Type:   typ,
+		Code:   code,
+		Length: len(icmp4.BaseLayer.Payload),
+	}
+	icmp.processMessage(tuple, &msg)
+}
+
+func (icmp *Icmp) ProcessICMPv6(icmp6 *layers.ICMPv6, pkt *protos.Packet) {
+	typ := uint8(icmp6.TypeCode >> 8)
+	code := uint8(icmp6.TypeCode)
+	id, seq := extractTrackingData(6, typ, &icmp6.BaseLayer)
+	tuple := NewIcmpTuple(6, pkt.Tuple.Src_ip, pkt.Tuple.Dst_ip, id, seq)
+	msg := IcmpMessage{
+		Ts:     pkt.Ts,
+		Type:   typ,
+		Code:   code,
+		Length: len(icmp6.BaseLayer.Payload),
+	}
+	icmp.processMessage(tuple, &msg)
+}
+
+func (icmp *Icmp) processMessage(tuple *IcmpTuple, msg *IcmpMessage) {
+	if isRequest(tuple, msg) {
+		icmp.processRequest(tuple, msg)
+	} else {
+		icmp.processResponse(tuple, msg)
+	}
+}
+
+func (icmp *Icmp) processRequest(tuple *IcmpTuple, msg *IcmpMessage) {
+	logp.Debug("icmp", "Processing request. %s", tuple)
+
+	trans := icmp.deleteTransaction(tuple.Hashable())
+	if trans != nil {
+		trans.Notes = append(trans.Notes, DuplicateRequestMsg)
+		logp.Debug("icmp", DuplicateRequestMsg+" %s", tuple)
+		icmp.publishTransaction(trans)
+	}
+
+	trans = newTransaction(msg.Ts, *tuple)
+	trans.Request = msg
+
+	if requiresCounterpart(tuple, msg) {
+		icmp.transactions.Put(tuple.Hashable(), trans)
+	} else {
+		icmp.publishTransaction(trans)
+	}
+}
+
+func (icmp *Icmp) processResponse(tuple *IcmpTuple, msg *IcmpMessage) {
+	logp.Debug("icmp", "Processing response. %s", tuple)
+
+	revTuple := tuple.Reverse()
+	trans := icmp.deleteTransaction(revTuple.Hashable())
+	if trans == nil {
+		trans = newTransaction(msg.Ts, revTuple)
+		trans.Notes = append(trans.Notes, OrphanedResponseMsg)
+		logp.Debug("icmp", OrphanedResponseMsg+" %s", tuple)
+	}
+
+	trans.Response = msg
+	icmp.publishTransaction(trans)
+}
+
+func (icmp *Icmp) direction(t *IcmpTransaction) uint8 {
+	if !icmp.isLocalIp(t.Tuple.SrcIp) {
+		return IcmpDirectionFromOutside
+	}
+	if !icmp.isLocalIp(t.Tuple.DstIp) {
+		return IcmpDirectionFromInside
+	}
+	return IcmpDirectionLocalOnly
```

Comment:
- add option (with support for netmask) to configure additional 'local' ips. So we can distinguish between adresses from inside/outside of network.
- fallback: use "bytes_out" for response and "bytes_in" for request (seems to be most consistent).